src/runtime/relax_vm/audio_support.cc

```diff
@@ -0,0 +1,332 @@
+    double matmul_result = 0.0;
+    for (int k = 0; k < n_fft / 2 + 1; k++) {
+      matmul_result += fft_out[k] * mel_filters[i][k];
+    }
+    matmul_result = std::max(matmul_result, 1e-10);
+    matmul_result = std::log10(matmul_result);
+    log_specs[num_id * n_mels + i] = matmul_result;
+  }
+}
+
+NDArray WhisperProcessAudio(NDArray raw_speech) {
+  ICHECK(raw_speech.IsContiguous());
+  ICHECK(raw_speech.DataType() == DataType::Float(32)) << "raw speech data type is not float32!";
+  ICHECK(raw_speech->device.device_type == kDLCPU) << "raw speech device must be CPU!";
+  ICHECK_EQ(raw_speech->ndim, 1);
+
+  const float* p_data = static_cast<float*>(raw_speech->data);
+
+  int sampling_rate = 16000;
+  int n_fft = 400;
+  int n_mels = 80;
+  int max_length = 480000;
+  int hop_length = 160;
+
+  std::vector<std::vector<double>> mel_filters = get_mel_filters(sampling_rate, n_fft, n_mels);
+
+  std::vector<double> window = hanning_window(n_fft + 1, n_fft);
+
+  std::vector<double> pad_data;
+  pad_data.resize(max_length + n_fft);
+  for (int i = 0; i < n_fft / 2; ++i) {
+    pad_data[n_fft / 2 - 1 - i] = p_data[i + 1];
+  }
+  for (int i = 0; i < max_length; ++i) {
+    if (i >= raw_speech->shape[0]) {
+      pad_data[n_fft / 2 + i] = 0.0;
+    } else {
+      pad_data[n_fft / 2 + i] = p_data[i];
+    }
+  }
+  for (int i = 0; i < n_fft / 2; ++i) {
+    pad_data[n_fft / 2 + max_length + i] = p_data[max_length - 2 - i];
+  }
+
+  int num_frames = 1 + (pad_data.size() - n_fft) / hop_length;
+  double* log_specs;
+  int output_num_frames = (num_frames - 1);
+  log_specs = new double[output_num_frames * n_mels];
+  for (int i = 0; i < output_num_frames; ++i) {
+    log_mel_spec(pad_data, i, window, n_fft, hop_length, mel_filters, log_specs);
+  }
+
+  double log_specs_max = std::numeric_limits<float>::min();
+  for (int i = 0; i < output_num_frames; i++) {
+    for (int j = 0; j < n_mels; ++j) {
+      log_specs_max = std::max(log_specs_max, log_specs[i * n_mels + j]);
+    }
+  }
+
+  log_specs_max -= 8.0;
+  for (int i = 0; i < output_num_frames; i++) {
+    for (int j = 0; j < n_mels; ++j) {
+      log_specs[i * n_mels + j] = std::max(log_specs_max, log_specs[i * n_mels + j]);
+      log_specs[i * n_mels + j] = (log_specs[i * n_mels + j] + 4.0) / 4.0;
+    }
+  }
+
+  auto ret_value = runtime::NDArray::Empty({1, n_mels, output_num_frames}, DataType::Float(32),
+                                           DLDevice{kDLCPU, 0});
+  float* p_ret = static_cast<float*>(ret_value->data);
+  for (int i = 0; i < output_num_frames; i++) {
+    for (int j = 0; j < n_mels; ++j) {
+      p_ret[j * output_num_frames + i] = static_cast<float>(log_specs[i * n_mels + j]);
+    }
+  }
+  return ret_value;
+}
+
+TVM_REGISTER_GLOBAL("vm.builtin.whisper_process_audio").set_body_typed(WhisperProcessAudio);
+
+// This is an inplace operation.
+void WhisperProcessLogits(NDArray logits, double cur_len) {
+  std::vector<int> suppress_tokens = {
```

Comment:
Can we allocate it outside of the function (in global memory space instead of stack)?